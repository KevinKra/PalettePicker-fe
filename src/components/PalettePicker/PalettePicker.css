.PalettePicker {
  border: 1px solid red;
  height: 70vh;
}
