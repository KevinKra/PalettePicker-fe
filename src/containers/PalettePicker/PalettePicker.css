.PalettePicker {
  display: flex;
  flex-direction: column;
  position: relative;
  height: 70vh;
  overflow: hidden;
  z-index: 2;
  border-bottom: 10px solid coral;
  box-shadow: -1px 10px 12px 5px rgba(0, 0, 0, 0.75);
}

.colors-section {
  display: flex;
  flex: 1;
  flex-direction: row;
}

.button-bar {
  position: relative;
  justify-content: center;
  align-items: center;
  display: flex;
  height: 15%;
  z-index: 2;
  border-top: 1px solid black;
  background: white;
}

.primary-btn {
  border: 1px solid transparent;
  background-color: coral;
  height: 50%;
  width: 45%;
  margin: 0.5rem;
  color: white;
  font-size: 1rem;
  border-radius: 5px;
}

.locked-btn {
  cursor: auto !important;
  background-color: lightgray;
}

button:focus {
  outline: none;
  cursor: pointer;
}
