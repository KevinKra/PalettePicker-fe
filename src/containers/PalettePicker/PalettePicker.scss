@import "../../_styles/main.scss";

.PalettePicker {
	grid-row: 1 / 2;
	display: flex;
	flex-direction: column;
	position: relative;
	overflow: hidden;
	z-index: 2;
	border-bottom: 10px solid $primary-color;
	box-shadow: -1px 10px 12px 5px rgba(0, 0, 0, 0.75);
	.colors-section {
		display: flex;
		flex: 1;
		flex-direction: row;
	}
	.button-bar {
		position: relative;
		justify-content: center;
		align-items: center;
		display: flex;
		height: 15%;
		z-index: 2;
		border-top: 1px solid black;
		background: white;
	}
	.primary-btn {
		@include btn-style-1;
		height: 50%;
		width: 45%;
		margin: 0.5rem;
		font-size: 16px;
	}
}
