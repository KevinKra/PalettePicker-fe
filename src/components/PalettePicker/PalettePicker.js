import React, { Component } from "react";
import ColorScheme from "color-scheme";
import uuidv1 from "uuid/v1";
import ColorBar from "../ColorBar/ColorBar";
import "./PalettePicker.css";

export default class PalettePicker extends Component {
  state = {
    hue: null,
    hueLocked: false,
    colorScheme: "triade",
    variation: "pastel",
    colors: [],
    editable: false
  };

  componentDidMount() {
    this.generateColors();
  }

  toggleEditable = () => {
    const toggle = this.state.editable;
    this.setState({ editable: !toggle });
  };

  updateHue = e => {
    const { value } = e.target;
    this.setState({ hue: value });
  };

  hueLock = () => {
    this.setState({ hueLocked: true });
  };

  hueUnlock = () => {
    this.setState({ hueLocked: false });
  };

  updateColorScheme = e => {
    const { value } = e.target;
    console.log(value);
    this.setState({ colorScheme: value });
  };

  updateVariation = e => {
    const { value } = e.target;
    this.setState({ variation: value });
  };

  updateColors = (e, previousColors) => {
    e.preventDefault();
    this.generateColors(previousColors);
  };

  generateRandomHue = () => {
    return Math.floor(Math.random() * (360 + 1));
  };

  handleLockStatus = (targetColor, lockStatus) => {
    const colorIndex = this.state.colors.findIndex(color => {
      return targetColor === color;
    });
    const colors = this.state.colors.slice();
    colors[colorIndex].locked = lockStatus;
    this.setState({ colors });
  };

  generateColors = async (previousColors = []) => {
    // The possible values are 'mono', 'contrast', 'triade', 'tetrade', and 'analogic'
    const { hue, hueLocked, colorScheme, variation, colors } = this.state;
    const { pColorScheme = "triade" } = this.props;
    let generatedHue;
    if (!hueLocked) {
      generatedHue = this.generateRandomHue();
    }
    const scheme = new ColorScheme();
    scheme
      .from_hue(generatedHue || hue)
      .scheme(colorScheme || pColorScheme)
      .variation(variation);
    const generatedColors = scheme.colors().map(color => {
      return { hex: "#" + color, locked: false };
    });
    if (previousColors.length !== generatedColors.length) {
      this.setState({ colors: generatedColors, hue: generatedHue || hue });
    } else {
      colors.forEach((color, i) => {
        if (color.locked === true) generatedColors.splice(i, 1, color);
      });
      this.setState({ colors: generatedColors, hue: generatedHue || hue });
    }
  };

  render() {
    const editBarActive = {
      transform: "translateY(0%)"
    };
    const colors = this.state.colors.map((color, i) => {
      return (
        <ColorBar
          color={color}
          vRotate={this.props.vRotate}
          number={i}
          handleLockStatus={this.handleLockStatus}
          // key={uuid}
        />
      );
    });

    const PhraseBlock = (
      <section
        className={`${
          !this.state.editable
            ? "PhraseBlock pb-active"
            : "PhraseBlock pd-inactive"
        }`}
      >
        <header className="phrase-background">
          <h2>Choose a color!</h2>
        </header>
        <section className="phrase-block-content">
          <button
            className={
              this.state.hueLocked
                ? "phrase-button locked-btn"
                : "phrase-button"
            }
            onClick={e => this.updateColors(e, this.state.colors)}
          >
            Refresh Colors
          </button>
          <div className="current-format">
            <div className="format-hue">
              <p>{this.state.hue || this.props.hue}</p>
              {this.state.hueLocked ? (
                <i className="fas fa-sm fa-lock" onClick={this.hueUnlock} />
              ) : (
                <i className="fas fa-sm fa-unlock-alt" onClick={this.hueLock} />
              )}
            </div>
            <p>{this.state.colorScheme || this.props.pColorScheme}</p>
            <p>{this.state.variation}</p>
          </div>
        </section>
      </section>
    );

    return (
      <section className="PalettePicker">
        {PhraseBlock}
        <div
          className="edit-block"
          style={this.state.editable ? editBarActive : null}
        >
          <form
            className="edits-form"
            onSubmit={e => this.updateColors(e, this.state.colors)}
          >
            <div className="hue-selection-container">
              <h4>Hue Selection:</h4>
              <div className="hue-selection-inputs">
                <label htmlFor="hue-selection">
                  <input
                    type="text"
                    placeholder="default: random"
                    autoComplete="off"
                    name="hue-selection"
                    value={this.state.hue}
                    onChange={this.updateHue}
                  />
                </label>
                {this.state.hueLocked ? (
                  <button onClick={this.hueUnlock}>Unlock</button>
                ) : (
                  <button onClick={this.hueLock}>Lock</button>
                )}
              </div>
            </div>
            <section className="radio-styles">
              <h4>Color schemes:</h4>
              <label htmlFor="mono">
                mono
                <input
                  type="radio"
                  name="colorScheme-selection"
                  value="mono"
                  onClick={e => this.updateColorScheme(e)}
                />
              </label>
              <label htmlFor="contrast">
                contrast
                <input
                  type="radio"
                  name="colorScheme-selection"
                  value="contrast"
                  onClick={e => this.updateColorScheme(e)}
                />
              </label>
              <label htmlFor="triade">
                triade
                <input
                  type="radio"
                  name="colorScheme-selection"
                  value="triade"
                  defaultChecked
                  onClick={e => this.updateColorScheme(e)}
                />
              </label>
              <label htmlFor="tetrade">
                tetrade
                <input
                  type="radio"
                  name="colorScheme-selection"
                  value="tetrade"
                  onClick={e => this.updateColorScheme(e)}
                />
              </label>
              <label htmlFor="analogic">
                analogic
                <input
                  type="radio"
                  name="colorScheme-selection"
                  value="analogic"
                  onClick={e => this.updateColorScheme(e)}
                />
              </label>
            </section>
            <section className="radio-styles">
              <h4>Color Variations:</h4>
              <label htmlFor="default">
                default
                <input
                  type="radio"
                  name="variation-selection"
                  value="default"
                  onClick={e => this.updateVariation(e)}
                />
              </label>
              <label htmlFor="pastel">
                pastel
                <input
                  type="radio"
                  name="variation-selection"
                  value="pastel"
                  defaultChecked
                  onClick={e => this.updateVariation(e)}
                />
              </label>
              <label htmlFor="soft">
                soft
                <input
                  type="radio"
                  name="variation-selection"
                  value="soft"
                  onClick={e => this.updateVariation(e)}
                />
              </label>
              <label htmlFor="light">
                light
                <input
                  type="radio"
                  name="variation-selection"
                  value="light"
                  onClick={e => this.updateVariation(e)}
                />
              </label>
              <label htmlFor="hard">
                hard
                <input
                  type="radio"
                  name="variation-selection"
                  value="hard"
                  onClick={e => this.updateVariation(e)}
                />
              </label>
              <label htmlFor="pale">
                pale
                <input
                  type="radio"
                  name="variation-selection"
                  value="pale"
                  onClick={e => this.updateVariation(e)}
                />
              </label>
            </section>
            <button className="update-btn">
              <p>Update</p>
            </button>
          </form>
        </div>
        <div className="colors-section">{colors}</div>
        <div className="button-bar">
          <button className="primary-btn" onClick={this.toggleEditable}>
            Edit
          </button>
          <button className="primary-btn">Save</button>
        </div>
      </section>
    );
  }
}
