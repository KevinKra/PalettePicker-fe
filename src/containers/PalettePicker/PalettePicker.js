import React, { Component } from "react";
import ColorScheme from "color-scheme";
import ColorBar from "../../components/ColorBar/ColorBar";
import EditBarFull from "../../components/EditBarFull/EditBarFull";
import EditBarPartial from "../../components/EditBarPartial/EditBarPartial";
import { withRouter } from "react-router-dom";
import * as actions from "../../_redux/actions";
import { connect } from "react-redux";
import "./PalettePicker.css";

class PalettePicker extends Component {
  state = {
    hue: "",
    hueLocked: false,
    colorScheme: "triade",
    variation: "pastel",
    colors: [],
    editable: false
  };

  componentDidMount() {
    this.generateColors();
  }

  componentDidUpdate() {
    this.props.updateCurrentPalette(this.state.colors);
  }

  saveDialogOpen = () => {
    this.props.history.push("save-palette");
  };

  toggleEditable = () => {
    const toggle = this.state.editable;
    this.setState({ editable: !toggle });
  };

  updateHue = e => {
    const { value } = e.target;
    this.setState({ hue: value });
  };

  hueLock = () => {
    this.setState({ hueLocked: true });
  };

  hueUnlock = () => {
    this.setState({ hueLocked: false });
  };

  updateColorScheme = e => {
    const { value } = e.target;
    console.log(value);
    this.setState({ colorScheme: value });
  };

  updateVariation = e => {
    const { value } = e.target;
    this.setState({ variation: value });
  };

  updateColors = (e, previousColors) => {
    e.preventDefault();
    this.generateColors(previousColors);
  };

  generateRandomHue = () => {
    return Math.floor(Math.random() * (360 + 1));
  };

  handleLockStatus = (targetColor, lockStatus) => {
    const colorIndex = this.state.colors.findIndex(color => {
      return targetColor === color;
    });
    const colors = this.state.colors.slice();
    colors[colorIndex].locked = lockStatus;
    this.setState({ colors });
  };

  generateColors = async (previousColors = []) => {
    // The possible values are 'mono', 'contrast', 'triade', 'tetrade', and 'analogic'
    const { hue, hueLocked, colorScheme, variation, colors } = this.state;
    const { pColorScheme = "triade" } = this.props;
    let generatedHue;
    if (!hueLocked) {
      generatedHue = this.generateRandomHue();
    }
    const scheme = new ColorScheme();
    scheme
      .from_hue(generatedHue || hue)
      .scheme(colorScheme || pColorScheme)
      .variation(variation);
    const generatedColors = scheme.colors().map(color => {
      return { hex: "#" + color, locked: false };
    });
    if (previousColors.length !== generatedColors.length) {
      this.setState({ colors: generatedColors, hue: generatedHue || hue });
    } else {
      colors.forEach((color, i) => {
        if (color.locked === true) generatedColors.splice(i, 1, color);
      });
      this.setState({ colors: generatedColors, hue: generatedHue || hue });
    }
  };

  render() {
    const colors = this.state.colors.map((color, i) => {
      return (
        <ColorBar
          color={color}
          vRotate={this.props.vRotate}
          number={i}
          handleLockStatus={this.handleLockStatus}
          key={i}
        />
      );
    });

    return (
      <section className="PalettePicker">
        <EditBarPartial
          editable={this.state.editable}
          hue={this.state.hue}
          hueLocked={this.state.hueLocked}
          colorScheme={this.state.colorScheme}
          variation={this.state.variation}
          colors={this.state.colors}
          updateColors={this.updateColors}
          hueLock={this.hueLock}
          hueUnlock={this.hueUnlock}
        />
        <EditBarFull
          editable={this.state.editable}
          hue={this.state.hue}
          hueLocked={this.state.hueLocked}
          colors={this.state.colors}
          updateColors={this.updateColors}
          updateHue={this.updateHue}
          hueLock={this.hueLock}
          hueUnlock={this.hueUnlock}
          updateVariation={this.updateVariation}
          updateColorScheme={this.updateColorScheme}
        />
        <div className="colors-section">{colors}</div>
        <div className="button-bar">
          <button className="primary-btn" onClick={this.toggleEditable}>
            Edit
          </button>
          <button className="primary-btn" onClick={this.saveDialogOpen}>
            Save
          </button>
        </div>
      </section>
    );
  }
}

//handle save button disable, needs to check elements that are locked.
// const mapStateToProps = ({

// })

const mapDispatchToProps = dispatch => ({
  createNewProject: project => dispatch(actions.createNewProject(project)),
  updateExistingProject: project =>
    dispatch(actions.updateExistingProject(project)),
  updateCurrentPalette: palette =>
    dispatch(actions.updateCurrentPalette(palette))
});

export default connect(
  null,
  mapDispatchToProps
)(withRouter(PalettePicker));
