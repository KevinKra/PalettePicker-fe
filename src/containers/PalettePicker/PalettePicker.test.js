import React from "react";
import { PalettePicker } from "./PalettePicker";
import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";

describe("PalettePicker", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PalettePicker />);
    const PalettePickerInstance = wrapper.instance();
  });
  it("should match the snapshot", () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  xit("componentDidUpdate: should invoke generateColors", () => {
    PalettePickerInstance.generateColors();
    expect(PalettePickerInstance.state.colors).toBe(true);
  });
});
