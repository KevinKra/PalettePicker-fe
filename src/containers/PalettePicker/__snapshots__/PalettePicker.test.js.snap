// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PalettePicker should match the snapshot 1`] = `
<section
  className="PalettePicker"
>
  <EditBarPartial
    colorScheme="triade"
    colors={
      Array [
        Object {
          "hex": "#71cc66",
          "locked": false,
        },
        Object {
          "hex": "#468040",
          "locked": false,
        },
        Object {
          "hex": "#d1e6cf",
          "locked": false,
        },
        Object {
          "hex": "#3fbf30",
          "locked": false,
        },
        Object {
          "hex": "#a35197",
          "locked": false,
        },
        Object {
          "hex": "#804076",
          "locked": false,
        },
        Object {
          "hex": "#e6cfe2",
          "locked": false,
        },
        Object {
          "hex": "#bf30aa",
          "locked": false,
        },
        Object {
          "hex": "#e69673",
          "locked": false,
        },
        Object {
          "hex": "#805340",
          "locked": false,
        },
        Object {
          "hex": "#e6d6cf",
          "locked": false,
        },
        Object {
          "hex": "#bf5c30",
          "locked": false,
        },
      ]
    }
    editable={false}
    hue={173}
    hueLock={[Function]}
    hueLocked={false}
    hueUnlock={[Function]}
    updateColors={[Function]}
    variation="pastel"
  />
  <EditBarFull
    colors={
      Array [
        Object {
          "hex": "#71cc66",
          "locked": false,
        },
        Object {
          "hex": "#468040",
          "locked": false,
        },
        Object {
          "hex": "#d1e6cf",
          "locked": false,
        },
        Object {
          "hex": "#3fbf30",
          "locked": false,
        },
        Object {
          "hex": "#a35197",
          "locked": false,
        },
        Object {
          "hex": "#804076",
          "locked": false,
        },
        Object {
          "hex": "#e6cfe2",
          "locked": false,
        },
        Object {
          "hex": "#bf30aa",
          "locked": false,
        },
        Object {
          "hex": "#e69673",
          "locked": false,
        },
        Object {
          "hex": "#805340",
          "locked": false,
        },
        Object {
          "hex": "#e6d6cf",
          "locked": false,
        },
        Object {
          "hex": "#bf5c30",
          "locked": false,
        },
      ]
    }
    editable={false}
    hue={173}
    hueLock={[Function]}
    hueLocked={false}
    hueUnlock={[Function]}
    updateColorScheme={[Function]}
    updateColors={[Function]}
    updateHue={[Function]}
    updateVariation={[Function]}
  />
  <div
    className="colors-section"
  >
    <ColorBar
      color={
        Object {
          "hex": "#71cc66",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="0"
      number={0}
    />
    <ColorBar
      color={
        Object {
          "hex": "#468040",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="1"
      number={1}
    />
    <ColorBar
      color={
        Object {
          "hex": "#d1e6cf",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="2"
      number={2}
    />
    <ColorBar
      color={
        Object {
          "hex": "#3fbf30",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="3"
      number={3}
    />
    <ColorBar
      color={
        Object {
          "hex": "#a35197",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="4"
      number={4}
    />
    <ColorBar
      color={
        Object {
          "hex": "#804076",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="5"
      number={5}
    />
    <ColorBar
      color={
        Object {
          "hex": "#e6cfe2",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="6"
      number={6}
    />
    <ColorBar
      color={
        Object {
          "hex": "#bf30aa",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="7"
      number={7}
    />
    <ColorBar
      color={
        Object {
          "hex": "#e69673",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="8"
      number={8}
    />
    <ColorBar
      color={
        Object {
          "hex": "#805340",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="9"
      number={9}
    />
    <ColorBar
      color={
        Object {
          "hex": "#e6d6cf",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="10"
      number={10}
    />
    <ColorBar
      color={
        Object {
          "hex": "#bf5c30",
          "locked": false,
        }
      }
      handleLockStatus={[Function]}
      key="11"
      number={11}
    />
  </div>
  <div
    className="button-bar"
  >
    <button
      className="primary-btn"
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      className="primary-btn"
      onClick={[Function]}
    >
      Save
    </button>
  </div>
</section>
`;
